### PR TITLE
SCA: Respect the "capture charge immediately" setting when using Payment Intents

### LINKED ISSUE #830: SCA: Respect the "capture charge immediately" setting when using Payment Intents
By default, a Payment Intent is created with `capture_method=automatic` ([API docs](https://stripe.com/docs/api/payment_intents/create#create_payment_intent-capture_method)). That means that, when the payment is accepted, it's authorized + captured in the same step.

In the Stripe gateway settings, there's an option labeled `Capture charge immediately`, that can be disabled. When disabled, the charge should just be authorized, and the merchant will be able to capture it later by changing the order's state, or by capturing the charge directly in the Stripe dashboard. We should respect that same behaviour for Payment Intents too.

Tasks:
- Allow capturing the intent when changing the order status, while preserving support for legacy charges.
- Listen to webhooks for when the merchant captures the Intent in the Stripe dashboard.
- Listen to webhooks that are triggered when the Intent is authorized but not captured.
- Deal with refunds correctly when the payment has just been authorized (not captured).

### PR DESCRIPTION
**Pull request by DanReyLop**
Fixes #830 

I had to do a small refactor, to move the intent-related utility functions from `WC_Gateway_Stripe` to the parent class, `WC_Stripe_Payment_Gateway`. `WC_Stripe_Order_Handler` extends `WC_Stripe_Payment_Gateway` and I needed to call those functions from there. Really, only `get_intent_from_order()` was needed but it made no sense to move only that function.

I'm branching of the `sca/update-api` branch to avoid merge conflicts.

Implementation note: I left the existing logic untouched as much as possible, opting instead for tweaking the `payment_intents/:id:/capture` and `payment_intents/:id/cancel` responses to conform to what the existing logic expects (a `charge` object and a `refund` object, respectively). We shouldn't follow this approach for WooPayments, since we'll use Payment Intents from the start.

## To test
Disable the "Capture charge immediately" setting, and then purchase items until you have enough test orders. These are the cases I personally tested:
- Transition an order from "on-hold" to "processing" (will trigger capture)
- Transition an order from "on-hold" to "cancelled" (will trigger full refund)
- Capture the Payment on Stripe Dashboard (will update the order notes after the webhook fires)
- Cancel the Payment on Stripe Dashboard (will update the order notes after the webhook fires)

### CODE REVIEW

File: includes/class-wc-stripe-order-handler.php [hunk 1]
```diff
@@ -237,29 +237,58 @@ public function capture_payment( $order_id ) {
...
+							),
+							'payment_intents/' . $intent->id . '/capture'
+						);
+
+						if ( ! empty( $result->error ) ) {
+							/* translators: error message */
+							$order->update_status( 'failed', sprintf( __( 'Unable to capture charge! %s', 'woocommerce-gateway-stripe' ), $result->error->message ) );
+						} else {
+							$is_stripe_captured = true;
+							$result = end( $intent->charges->data );
```
Comment (dechov) [inline includes/class-wc-stripe-order-handler.php]: Should this not be pulled from `$result` rather than `$intent`?

Comment (DanReyLop) [reply includes/class-wc-stripe-order-handler.php]: Yes, fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/836/commits/6d82f8c98c0525ed688d2d8f02029df5040652df, good catch!

**Comment by dechov** [comment]
@DanReyLop Not sure if it gives a clue, but to follow up on that 3D Secure test, I seemed to get no error logs when it stalled and didn't redirect, but the next time I tried, I saw this log output:

```
[11-Apr-2019 19:35:45 UTC] PHP Notice:  Undefined property: stdClass::$status in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php on line 1138
[11-Apr-2019 19:35:45 UTC] PHP Warning:  end() expects parameter 1 to be array, null given in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php on line 842
[11-Apr-2019 19:35:45 UTC] PHP Notice:  Undefined property: stdClass::$status in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-gateway-stripe.php on line 845
[11-Apr-2019 19:35:45 UTC] PHP Notice:  Trying to get property 'id' of non-object in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php on line 473
[11-Apr-2019 19:35:45 UTC] PHP Notice:  Trying to get property 'id' of non-object in /srv/www/wordpress-default/public_html/wp-content/plugins/woocommerce-gateway-stripe/includes/abstracts/abstract-wc-stripe-payment-gateway.php on line 480
```

Also, it says "This branch has conflicts" – does the base just need to be changed to `sca/master`?

**Comment by dechov** [comment]
Opened https://github.com/woocommerce/woocommerce-gateway-stripe/issues/850, https://github.com/woocommerce/woocommerce-gateway-stripe/issues/851, https://github.com/woocommerce/woocommerce-gateway-stripe/issues/853, and https://github.com/woocommerce/woocommerce-gateway-stripe/issues/852 to track bugs mentioned (in that order) in previous comment, the first two of which are in milestone 4.2.

**Comment by DanReyLop** [comment]
@dechov Thanks for creating the issues! It makes total sense to only tackle the regressions in 4.2.0, thanks for making that distinction.

**Pull request merged**